File: java/src/leetcode/binary_tree/IsValidBST.java
package leetcode.binary_tree;

/**
 * @author chengzw
 * @description 验证二叉搜索树，https://leetcode-cn.com/problems/validate-binary-search-tree/
 * <p>
 * 给你一个二叉树的根节点 root ，判断其是否是一个有效的二叉搜索树。
 * <p>
 * 有效二叉搜索树定义如下：
 * 节点的左子树只包含 小于 当前节点的数。
 * 节点的右子树只包含 大于 当前节点的数。
 * 所有左子树和右子树自身必须也是二叉搜索树。
 * @since 2021/10/6
 */

/**
 * 思路：
 * 1.返回值并非要求解的值
 * 2.要求解的值放在全局变量中 isValid
 * 3.使用递归，如果左子树的最大值大于等于当前节点或者右子树的最小值小于等于当前节点，修改 isValid 为 false，然后提前归
 */
public class IsValidBST {
    //全局变量
    public boolean isValid = true;

    public boolean isValidBST(TreeNode root) {
        //特殊情况
        if (root == null) return true;
        dfs(root);
        return isValid;
    }

    public int[] dfs(TreeNode root) {
        int min = root.val;
        int max = root.val;

        if (root.left != null) {
            int[] leftMinMax = dfs(root.left);
            //提前返回
            if (isValid == false) return null;
            //如果左子树的最大值大于等于当前节点的值
            if (leftMinMax[1] >= root.val) {
                isValid = false;
                return null;
            }
            min = leftMinMax[0];
        }


        if (root.right != null) {
            int[] rightMinMax = dfs(root.right);
            //提前返回
            if (isValid == false) return null;
            //如果右子树的最小值小于等于当前节点的值
            if (rightMinMax[0] <= root.val) {
                isValid = false;
                return null;
            }
            max = rightMinMax[1];
        }
        return new int[]{min, max};
    }
}

File: go/leetcode/binary_tree/is_valid_bst.go
package binary_tree

/**
 * @description 98. 验证二叉搜索树
 * @author chengzw
 * @since 2022/6/13
 * @link https://leetcode.cn/problems/validate-binary-search-tree/
 */

/**
给你一个二叉树的根节点 root ，判断其是否是一个有效的二叉搜索树。
有效 二叉搜索树定义如下：
节点的左子树只包含 小于 当前节点的数。
节点的右子树只包含 大于 当前节点的数。
所有左子树和右子树自身必须也是二叉搜索树。

示例 1：
输入：root = [2,1,3]
输出：true

示例 2：
输入：root = [5,1,4,null,null,3,6]
输出：false
解释：根节点的值是 5 ，但是右子节点的值是 4 。
*/

/**
思路：
1.左子树的最大值是 root.val，右子树的最小值是 root.val。
2.左右子树同时满足 max 和 min 的限制说明是合法 BST。

时间复杂度：O(n)，其中 n 为二叉树的节点个数。在递归调用的时候二叉树的每个节点最多被访问一次，因此时间复杂度为 O(n)。
空间复杂度：O(n)，其中 n 为二叉树的节点个数。递归函数在递归过程中需要为每一层递归函数分配栈空间，
所以这里需要额外的空间且该空间取决于递归的深度，即二叉树的高度。
最坏情况下二叉树为一条链，树的高度为 n ，递归最深达到 n 层，故最坏情况下空间复杂度为 O(n) 。
*/
func isValidBST(root *TreeNode) bool {
	return helper2(root, nil, nil)
}

//使用辅助函数
func helper2(root *TreeNode, min *TreeNode, max *TreeNode) bool {
	//二叉树可以为空
	if root == nil {
		return true
	}
	// 若 root.val 不符合 max 和 min 的限制，说明不是合法 BST
	if min != nil && root.Val <= min.Val {
		return false
	}
	if max != nil && root.Val >= max.Val {
		return false
	}
	//左子树的最大值是 root.val，右子树的最小值是 root.val
	//左右子树同时满足 max 和 min 的限制说明是合法 BST
	return helper2(root.Left, min, root) && helper2(root.Right, root, max)
}
